packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -519,7 +519,9 @@ class _DatePickerDialogState extends State<DatePickerDialog> with RestorationMix
                   firstDate: widget.firstDate,
                   lastDate: widget.lastDate,
                   onDateSubmitted: _handleDateChanged,
-                  onDateSaved: _handleDateChanged,
+                  onDateSaved: (DateTime savedDate) {
+                    _selectedDate.value = savedDate;
```

Comment:
does the build method use _selectedDate?

Reply:
this should have a comment saying why we're not calling setState here.

In general it seems very brittle if calling setState breaks something.

Reply:
> does the build method use _selectedDate?

Whenever the form is validated this class is rebuilt as a result, `onDateSaved` with `_handleDateChanged` is called again when the user confirms the date input.  When the user confirms the valid date, the picker should just pop with the result, without setState at the end which duplicates `selectableDayPredicate` value.

full calendar picker mode doesn't have this issue since there is form validation there.

Reply:
Add a comment and here is the video we can see `onDateSaved` is called when a user submits and calling `_handleDateChanged` (with setState) will unnecessarily rebuild the dialog, and `selectableDayPredicate` is called before the dialog pops.

https://user-images.githubusercontent.com/48603081/180963351-38e7fe56-80dc-4e64-be45-2a4365331334.mov